aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -13,82 +13,146 @@ namespace native {
+}
+
+static TensorIterator make_index_iterator(const Tensor input, const std::vector<Tensor> indices) {
+  TensorIteratorConfig config;
+
+  auto output_tensor = Tensor();
+  if(input.is_quantized()) {
+    double scale = input.q_scale();
+    int64_t zero_point = input.q_zero_point();
+    output_tensor = at::_empty_affine_quantized(
+        input.sizes(), at::device(c10::kCPU).dtype(input.scalar_type()), scale, zero_point);
+  }
+
+  config.set_check_mem_overlap(false)
+        .check_all_same_dtype(false)
+        .declare_static_dtype_and_device(input.scalar_type(), input.device())
+        .add_output(output_tensor)
+        .add_input(input);
+  for (auto& index : indices) {
+    config.add_input(index);
+  }
+  return config.build();
+}
+
+struct Indexer {
+  Indexer(int64_t num_indexers, char** indexers, const int64_t* indexer_strides)
+      : num_indexers(num_indexers),
+        indexers(indexers),
+        indexer_strides(indexer_strides) {}
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = *(int64_t*)&indexers[0][idx * indexer_strides[0]];
+    for (int j = 1; j < num_indexers; j++) {
+      offset += *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+    }
+    return offset;
+  }
+};
+
 template <typename scalar_t>
-void inline flip_cpu_kernel(
-  const int64_t total_dims,
-  const std::vector<int64_t>& stride_contiguous_v,
-  const std::bitset<dim_bitset_size>& flip_dims_b,
-  const Tensor& in_tensor,
-  Tensor& out_tensor
-){
-  const int64_t numel = in_tensor.numel();
-  const scalar_t* in_tensor_d = in_tensor.data_ptr<scalar_t>();
-  scalar_t* out_tensor_d = out_tensor.data_ptr<scalar_t>();
-  auto sizes_v = in_tensor.sizes().vec();
-  auto strides_v = in_tensor.strides().vec();
-
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
-  at::parallel_for(0, numel, 1000, [&](int64_t start, int64_t end) {
-    for (auto i = start; i < end; i++) {
-      int64_t cur_indices = i;
-      int64_t rem = 0;
-      int64_t dst_offset = 0;
-
-      for (int64_t d = 0; d < total_dims; d++) {
-        int64_t temp = cur_indices;
-        cur_indices = cur_indices / stride_contiguous_v[d];
-        rem = temp - cur_indices * stride_contiguous_v[d];
-        dst_offset += flip_dims_b[d] ? (sizes_v[d] - 1 - cur_indices) * strides_v[d] : cur_indices * strides_v[d];
-        cur_indices = rem;
-      }
-      out_tensor_d[i] = in_tensor_d[dst_offset];
+void flip_cpu_kernel(TensorIterator& iter) {
+  int ntensor = iter.ntensors();
+  // When launch the index parallel version, set a relative samll grain size
+  // less than the INTERNAL::GRAIN_SIZE to make the whole available thread
+  // numbers get more balanced work load and a better cache location. The grain
+  // size here is chosen by the op benchmark to overcome the thread launch
+  // overhead. This value was taken from the AdvancedIndexing kernel.
+  const int index_parallel_grain_size = 3000;
```

Comment:
@andfoy any gains using this value vs default one ?

Reply:
Let me check!

Reply:
See https://github.com/pytorch/pytorch/pull/56713#issuecomment-837354845